corehq/util/couch.py

```diff
@@ -33,3 +36,146 @@ def categorize_bulk_save_errors(error):
         result_map[error].append(result)
 
     return result_map
+
+
+class IterDB(object):
+    """
+    Bulk save docs in chunks.
+
+        with IterDB(db) as iter_db:
+            for doc in iter_docs(db, ids):
+                iter_db.save(doc)
+
+        iter_db.error_ids  # docs that errored
+        iter_db.saved_ids  # docs that saved correctly
+    """
+    def __init__(self, database, chunksize=100, throttle_secs=None):
+        self.db = database
+        self.chunksize = chunksize
+        self.throttle_secs = throttle_secs
+        self.saved_ids = set()
+        self.deleted_ids = set()
+        self.error_ids = set()
+        self.errors_by_type = defaultdict(list)
+
+    def __enter__(self):
+        self.to_save = []
+        self.to_delete = []
+        return self
+
+    def _write(self, cmd, docs):
+        try:
+            results = cmd(docs)
+        except BulkSaveError as e:
+            categorized_errors = categorize_bulk_save_errors(e)
+            success_ids = {r['id'] for r in categorized_errors.pop(None, [])}
+            self.errors_by_type = categorized_errors
+            self.error_ids.update(d['id'] for d in e.errors)
+        else:
+            success_ids = {d['id'] for d in results}
+
+        if self.throttle_secs:
+            sleep(self.throttle_secs)
+        return success_ids
+
+    def _commit_save(self):
+        success_ids = self._write(self.db.bulk_save, self.to_save)
+        self.saved_ids.update(success_ids)
+        self.to_save = []
+
+    def _commit_delete(self):
+        success_ids = self._write(self.db.bulk_delete, self.to_delete)
+        self.deleted_ids.update(success_ids)
+        self.to_delete = []
+
+    def save(self, doc):
+        self.to_save.append(doc)
+        if len(self.to_save) >= self.chunksize:
+            self.commit_save()
+
+    def delete(self, doc):
+        self.to_delete.append(doc)
+        if len(self.to_delete) >= self.chunksize:
+            self.commit_delete()
+
+    def commit(self):
+        if self.to_save:
+            self.commit_save()
+        if self.to_delete:
+            self.commit_delete()
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.commit()
+
+
+class IterUpdateError(Exception):
+    def __init__(self, results, *args, **kwargs):
+        self.results = results
+        super(IterUpdateError, self).__init__(*args, **kwargs)
+
+
+def iter_update(db, fn, ids, max_retries=None):
```

Comment:
seems like it could be more readable to just set the default to 3 rather than doing it when it's accessed in the function